clash-lib/src/Clash/Rewrite/Util.hs

```diff
@@ -212,82 +194,87 @@ applyDebug
-                        ++ name ++ " to:\n" ++ after ++ "\n") $
-        return exprNew
+applyDebug name exprOld hasChanged exprNew = do
+  nTrans <- Lens.use transformCounter
+  opts <- Lens.view debugOpts
+
+  let from = fromMaybe 0 (dbg_transformationsFrom opts)
+  let limit = fromMaybe maxBound (dbg_transformationsLimit opts)
+
+  if | nTrans - from > limit ->
+         error "-fclash-debug-transformations-limit exceeded"
+     | nTrans <= from ->
+         pure exprNew
+     | otherwise ->
+         go opts
  where
-  before = showPpr exprOld
-  after  = showPpr exprNew
+  go opts = traceIf (hasDebugInfo TryTerm name opts) ("Tried: " ++ name ++ " on:\n" ++ before) $ do
+    nTrans <- pred <$> Lens.use transformCounter
+
+    Monad.when (dbg_countTransformations opts && hasChanged) $ do
+      transformCounters %= HashMap.insertWith (const succ) (Text.pack name) 1
+
+    Monad.when (dbg_invariants opts && hasChanged) $ do
+      tcm                  <- Lens.view tcCache
+      let beforeTy          = termType tcm exprOld
+          beforeFV          = Lens.setOf freeLocalVars exprOld
+          afterTy           = termType tcm exprNew
+          afterFV           = Lens.setOf freeLocalVars exprNew
+          newFV             = not (afterFV `Set.isSubsetOf` beforeFV)
+          accidentalShadows = findAccidentialShadows exprNew
+
+      Monad.when newFV $
+              error ( concat [ $(curLoc)
+                             , "Error when applying rewrite ", name
+                             , " to:\n" , before
+                             , "\nResult:\n" ++ after ++ "\n"
+                             , "It introduces free variables."
+                             , "\nBefore: " ++ showPpr (Set.toList beforeFV)
+                             , "\nAfter: " ++ showPpr (Set.toList afterFV)
+                             ]
+                    )
+      Monad.when (not (null accidentalShadows)) $
+        error ( concat [ $(curLoc)
+                       , "Error when applying rewrite ", name
+                       , " to:\n" , before
+                       , "\nResult:\n" ++ after ++ "\n"
+                       , "It accidentally creates shadowing let/case-bindings:\n"
+                       , " ", showPpr accidentalShadows, "\n"
+                       , "This usually means that a transformation did not extend "
+                       , "or incorrectly extended its InScopeSet before applying a "
+                       , "substitution."
+                       ])
+
+      -- TODO This check should not have the `hasDebugInfo` call in it, as
+      -- setting dbg_invariants should be all that is necessary to check this.
+      -- However, currently this error is very fragile, as Clash currently does
+      -- not keep casts, so "illegally" changing between `Signal dom a` and `a`
+      -- will trigger this error for many designs.
+      --
+      -- This should be changed when #1064 (PR to keep casts in core) is merged.
```

Comment:
Apart from casts, wasn't this also triggered by [`separateArguments`](https://github.com/clash-lang/clash-compiler/blob/be7558062e2832c446b9dcf61be3852872b646bb/clash-lib/src/Clash/Normalize/Transformations/SeparateArgs.hs#L46)?

Reply:
The implicit params thing? Yeah, I didn't include that because by the time the signal cast PR is in it might not be an issue / there might be other small issues so I didn't want to get too into detail past the most immediate obstacle